File: Top_Interview_Questions/minStack.cpp
class Node {
    public:
        Node() : val{0}, min{0}, next{nullptr} {}
        Node(int val, int min) : val{val}, min{min}, next{nullptr} {}
        Node(int val, int min, Node *next) : val{val}, min{min}, next{next} {}

        int val;
        int min;
        Node *next;
};

class MinStack {
public:
    void push(int val) {
        if (!head) {
            head = new Node{val, val};
        } else {
            head = new Node{val, min(val, head->min), head};
        }
    }
    
    void pop() {
        head = head->next;
    }
    
    int top() {
        return head->val;
    }
    
    int getMin() {
        return head->min;
    }

private:
    Node *head;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: NeetCode_Roadmap/min_stack.py
class Node:
    def __init__(self, val: int, min_val: int, next_n):
        self.val = val
        self.min_val = min_val
        self.next_n = next_n

class MinStack:

    def __init__(self):
        self.head = None

    def push(self, val: int) -> None:
        if self.head is not None:
            self.head = Node(
                val, min(self.head.min_val, val),
                self.head
            )
        else:
            self.head = Node(val, val, None)

    def pop(self) -> None:
        self.head = self.head.next_n

    def top(self) -> int:
        return self.head.val

    def getMin(self) -> int:
        return self.head.min_val
